src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -459,13 +382,66 @@ private void DetachFromDocument(uint docCookie, string moniker)
             }
         }
 
+        private void SubscribeToRunningDocTableEvents()
+        {
+            _runningDocumentTableEventSink.OnRenameDocument += HandleRenameDocumentEvent;
+            _runningDocumentTableEventSink.OnReloadDocumentData += HandleReloadDocumentEvent;
+            _runningDocumentTableEventSink.OnInitializedDocument += HandleInitializedDocumentEvent;
+            _runningDocumentTableEventSink.OnCloseDocument += HandleCloseDocumentEvent;
+        }
+
+        private void UnSubscribeFromRunningDocTableEvents()
+        {
+            _runningDocumentTableEventSink.OnRenameDocument -= HandleRenameDocumentEvent;
+            _runningDocumentTableEventSink.OnReloadDocumentData -= HandleReloadDocumentEvent;
+            _runningDocumentTableEventSink.OnInitializedDocument -= HandleInitializedDocumentEvent;
+            _runningDocumentTableEventSink.OnCloseDocument -= HandleCloseDocumentEvent;
+        }
+
+        private void HandleRenameDocumentEvent(object sender, RunningDocumentTableRenamedEventArgs args)
+        {
+            // We want to consider this file to be added in one of two situations:
+            //
+            // 1) the old file already was a misc file, at which point we might just be doing a rename from
+            //    one name to another with the same extension
+            // 2) the old file was a different extension that we weren't tracking, which may have now changed
+            if (TryUntrackClosingDocument(args.DocCookie, args.OldMoniker) || TryGetLanguageInformation(args.OldMoniker) == null)
+            {
+                if (_runningDocumentTableEventSink.TryGetBuffer(args.DocCookie, out var buffer))
+                {
+                    // Add the new one, if appropriate.
+                    TrackOpenedDocument(args.DocCookie, args.Moniker, buffer);
+                }
+            }
+        }
+
+        private void HandleReloadDocumentEvent(object sender, RunningDocumentTableEventArgs args)
+        {
+            if (args.Moniker != null && TryGetLanguageInformation(args.Moniker) != null && !_docCookiesToProjectIdAndContainer.ContainsKey(args.DocCookie))
+            {
+                if (_runningDocumentTableEventSink.TryGetBuffer(args.DocCookie, out var buffer))
+                {
+                    TrackOpenedDocument(args.DocCookie, args.Moniker, buffer);
+                }
+            }
+        }
+
+        private void HandleInitializedDocumentEvent(object sender, RunningDocumentTableInitializedEventArgs args)
+        {
+            // The document is now initialized, we should try tracking it
+            TrackOpenedDocument(args.DocCookie, args.Moniker, args.TextBuffer);
+        }
+
+        private void HandleCloseDocumentEvent(object sender, RunningDocumentTableEventArgs args)
+        {
+            _foregroundThreadAffinitization.AssertIsForeground();
```

Comment:
reason only this has assert?

Reply:
Copied from the original - on the close event there is an assert.  On other events it's handled inside the trackopendocument method.  But I believe since tryuntrackclosed document is used in multiple cases, the assert was left outside